### PR TITLE
add prepare_pretrained_model

### PR DESCRIPTION
**Pull request by yuyu2172**
I think it is a good time to make a utility function that handles `n_class,n_fg_class`, `pretrained_model` and `models`.
These operations overlap between different implementations.
Also, this part is verbose and non-informative.

TODOs

~~Write tests with different initialization options including `pretrained_model` for some models.~~

### CODE REVIEW

File: chainercv/utils/link.py [hunk 1]
```diff
@@ -10,6 +10,29 @@
...
+            :obj:`model` has four keys: :obj:`param`, :obj:`overwritable`,
+            :obj:`url` and :obj:`cv2`.
+
+            * **param** (*dict*): Parameters assigned to the pretrained \
+                weight.
+            * **overwritable** (*set*): Names of parameters that are \
+                overwritable (i.e., :obj:`param[key] != model['param'][key]` \
+                is accepted).
+            * **url** (*string*): Location of the pretrained weight.
+            * **cv2** (*bool*): If :obj:`True`, an error is raised \
```
Comment (Hakuyume) [inline chainercv/utils/link.py]: error -> warning

**Comment by yuyu2172** [comment]
I supported the models to skip loading class dependent weights.

File: chainercv/links/model/yolo/yolo_v2.py [hunk 2]
```diff
@@ -150,7 +159,12 @@ def __init__(self, n_fg_class=None, pretrained_model=None):
         self._default_bbox = np.array(default_bbox, dtype=np.float32)
 
         if path:
-            chainer.serializers.load_npz(path, self, strict=False)
+            if use_pretrained_class_weights:
+                chainer.serializers.load_npz(path, self, strict=False)
+            else:
+                utils.link.load_npz_with_ignore_names(
+                    path, self, strict=False,
+                    ignore_names=['subnet/W', 'subnet/b'])
```
Comment (Hakuyume) [inline chainercv/links/model/yolo/yolo_v2.py]: `subnet` contains the weights for localization and objectness. This code removes them.

File: chainercv/links/model/yolo/yolo_v3.py [hunk 3]
```diff
@@ -169,7 +178,16 @@ def __init__(self, n_fg_class=None, pretrained_model=None):
...
 
         if path:
-            chainer.serializers.load_npz(path, self, strict=False)
+            if use_pretrained_class_weights:
+                chainer.serializers.load_npz(path, self, strict=False)
+            else:
+                class_dependent_weight_names = [
+                    'subnet/0/1/b', 'subnet/0/1/W',
+                    'subnet/1/1/b', 'subnet/1/1/W',
+                    'subnet/2/1/b', 'subnet/2/1/W']
```
Comment (Hakuyume) [inline chainercv/links/model/yolo/yolo_v3.py]: ditto.

**Comment by yuyu2172** [comment]
I think `class_weight` problem should be solved together with `prepare_pretrained_model`.

Sorry for bugs in YOLO.
How about removing these options for YOLO?

**Comment by yuyu2172** [comment]
Any name recommendation?

File: chainercv/utils/link.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        if param[key] is None:
+            if key in default:
+                param[key] = default[key]
+            else:
+                raise ValueError('{} must be specified'.format(key))
+
+    return param, path
+
+
+def load_npz_with_ignore_names(file, obj, path='', strict=True, ignore_names=None):
```
Comment (Hakuyume) [inline chainercv/utils/link.py]: It's time to use `chainer_experimental`.

**Comment by Hakuyume** [comment]
> How about removing these options for YOLO?

I think it is better to support YOLO. However, you can omit the support in this PR because it requires some modifications in YOLO.

**Comment by yuyu2172** [comment]
Summary of offline discussion:

- [ ] Delete `use_pretrained_class_weight` option
- [ ] Move `save_npy_ignore_names` to `chainer_experimental`
- [ ] Write a tutorial on how to cope with the problem `use_pretrained_class_weight` is trying to solve (write a method that takes the source and the target models and ignore names).

File: tests/links_tests/model_tests/vgg_tests/test_vgg16.py [hunk 5]
```diff
@@ -5,17 +5,17 @@
...
 from chainercv.links import VGG16
+from chainercv.testing import attr
 
 
 @testing.parameterize(
     {'pick': 'prob', 'shapes': (1, 200), 'n_class': 200},
     {'pick': 'pool5', 'shapes': (1, 512, 7, 7), 'n_class': None},
     {'pick': ['conv5_3', 'conv4_2'],
-     'shapes': ((1, 512, 14, 14), (1, 512, 28, 28)), 'n_class': None},
+     'shapes': ((1, 512, 14, 14), (1, 512, 28, 28)), 'n_class': None}
```
Comment (Hakuyume) [inline tests/links_tests/model_tests/vgg_tests/test_vgg16.py]: What is the intent of this modification?

File: docs/source/reference/utils.rst [hunk 6]
```diff
@@ -64,6 +64,14 @@ unzip
 .. autofunction:: unzip
 
 
+Link Utilities 
+---------------
```
Comment (Hakuyume) [inline docs/source/reference/utils.rst]: The length of `---`.

File: tests/links_tests/model_tests/vgg_tests/test_vgg16.py (hunk 5 above)
Comment (yuyu2172) [reply tests/links_tests/model_tests/vgg_tests/test_vgg16.py]: No specific intention.  I can put it back

**Pull request merged**